### DISCUSSION TITLE
packageRules and postUpdateOptions only for a part of a monorepo

### DISCUSSION THREAD

**Discussion by maximilianbraun** [renovatebot/renovate#13435] [Request Help] [answered]
Hi!

I'm looking for a way to have certain postUpdateOptions only running for parts of a monorepo. I would like to have this modules to opt out of the postUpdateOptions.

Our default config contains a preset for postUpgradeOptions, which i set to ignored. My last attempt was to use a glob with negation.

However, It didn't work out and all modules get their dedupe. 

Any tip / docs / example / suggestions? 

```json
 {
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "local>yxz/renovate-config",
    "local>yxz/renovate-config:npm",
    "local>yxz/renovate-config:golang"
  ],
  "ignorePresets": [
    "local>yxz/renovate-config:post-upgrade-options"
  ],
  "packageRules": [
    {
      "matchPaths": [
        "**/!(moduleA|moduleB|moduleC)*/**"
      ],
      "description": "no dedupe",
      "postUpdateOptions": [
        "npmDedupe"
      ]
    }
```

**Comment by rarkins** [comment, marked as answer]
I would have expected that `postUpdateOptions` within `packageRules` should work. Can you create a reproduction repo to demonstrate the failure so we can debug?

**Reply by maximilianbraun** [reply]
Hi @rarkins sure.
https://github.com/maximilianbraun/renovate-monorepo 
![image](https://user-images.githubusercontent.com/78038429/148798717-26b4c80f-7db0-4720-91a2-89b6b171d6f6.png)
![image](https://user-images.githubusercontent.com/78038429/148799704-83affb09-85cf-44fc-910d-62b32d57fd59.png)

**Reply by maximilianbraun** [reply]
I think I need to rephrase the problem: the post upgrade options are used, but for all the modules, even though they should be excluded by the glob pattern. Maybe i'm holding it wrong, not sure.

**Reply by maximilianbraun** [reply]
I updated the structure, to match a bit better the problem. Still i see dedupe executed in the logs.

What i tried was to add a test-case in `package-rules.spec.ts` which eventually passes. 
```
it('matches paths with negation', () => {
    const config: TestConfig = {
      packageFile: 'examples/foo/package.json',
      packageRules: [
        {
          matchPaths: ['*/!(moduleA|moduleD)/**'],
          x: 1,
        },
      ],
    };


    config.packageFile = 'subfolder0/moduleA/package.json';
    const moduleA = applyPackageRules({
      ...config,
      depName: 'test',
    });

    config.packageFile = 'subfolder0/moduleB/package.json';
    const moduleB = applyPackageRules({
      ...config,
      depName: 'test',
    });


    config.packageFile = 'subfolder0/moduleC/package.json';
    const moduleC = applyPackageRules({
      ...config,
      depName: 'test',
    });

    config.packageFile = 'subfolder1/moduleD/package.json';
    const moduleD = applyPackageRules({
      ...config,
      depName: 'test',
    });

    config.packageFile = 'subfolder1/moduleE/package.json';
    const moduleE = applyPackageRules({
      ...config,
      depName: 'test',
    });

    expect(moduleA.x).toBeUndefined();
    expect(moduleB.x).toBeDefined();
    expect(moduleC.x).toBeDefined();
    expect(moduleD.x).toBeUndefined();
    expect(moduleE.x).toBeDefined();
  });
 ```

**Reply by rarkins** [reply]
So it seems the problem may be with the `matchPaths` glob? Can you change your use case so that it's easier to test? e.g. instead of `postUpdateOptions`, use `"additionalBranchPrefix": "extra-"`. Then you can see if it's been applied by looking at the branch name and not needing to check the logs.

**Reply by maximilianbraun** [reply]
Done. Interestingly, it we do now have two PR's, one for A+D and one for the rest. Which is what we want.

So... the glob works but there is some merging of postUpdateOptions happening?

**Reply by maximilianbraun** [reply]
Short update:
I now used an allowlist for the postUpdateOptions. Still had the issue. I created another packageRule with a list of the other modules and just added an additionalBranchPrefix. Now it is, of course, split into two PRs and the handling of the postUpdateOptions works well.

**Reply by rarkins** [reply]
Post update options will intentionally be merged within one PR if there are multiple updates within that PR. Is that the case here?

**Reply by maximilianbraun** [reply]
looks like so. Ok, then i understood and got my solution.

Thanks @rarkins
